### PR TITLE
Throw exception when modifying contents of Container<T> while iterating through it

### PR DESCRIPTION
**Pull request by hwabis**
Closes https://github.com/ppy/osu-framework/issues/5379

I believe I've covered all the methods that can mutate `Children`?

Also not sure if I should also check the version in `Reset()` of the enumerator. `List<T>` does it but I can't think of the case where that would throw

### CODE REVIEW

File: osu.Framework/Graphics/Containers/Container.cs [hunk 1]
```diff
@@ -36,6 +36,13 @@ public class Container : Container<Drawable>
     public class Container<T> : CompositeDrawable, IContainerEnumerable<T>, IContainerCollection<T>, ICollection<T>, IReadOnlyList<T>
         where T : Drawable
     {
+        /// <summary>
+        /// This is checked when enumerating through this <see cref="Container"/> to throw when
+        /// <see cref="Children"/> was mutated while enumerating (in <see cref="Enumerator"/>).
+        /// This is incremented whenever <see cref="Children"/> is mutated (e.g. with <see cref="Add(T)"/>). 
+        /// </summary>
+        private int version = 0;
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Containers/Container.cs:44]: I think this should probably be named like `enumeratorVersion` instead, just to indicate its use.

**Review by smoogipoo — APPROVED** [review]

**Comment by peppy** [comment]
@smoogipoo did you happen to test this again osu-side tests? shouldn't stop making the change, but curious if we have any work to do there.

**Review by peppy — APPROVED** [review]

**Comment by smoogipoo** [comment]
Ah nope, I haven't checked

**Comment by smoogipoo** [comment]
No test failures.

**Comment by frenzibyte** [comment]
Will note that this doesn't cover the case where the contents of `Container<T>` is actually modified, it only handles addition/deletion. That being said, not 100% sure if it should cover that. Maybe best to do it in order to remain consistent with `List<T>`'s behaviour?

**Pull request merged**